docs/adr/000x-modelling-history.md

```diff
@@ -0,0 +1,250 @@
+model.
+
+### Core Concepts
+
+![Main concepts](000x/main-concepts-diagram.png)
+
+**Document**: A record that represents all versions of a piece of content in a
+particular locale. It has many editions and at any time it will have a current
+edition - shown on Content Publisher index - and potentially a live edition
+which is currently on GOV.UK. The live and current edition can be
+the same. Each iteration of a document's content is represented as a revision,
+thus a document has many revisions. Document is a mutable entity that tracks
+data across all editions (such as first publishing date) and it is expected to
+be a joining point for document-related data that is not associated with a
+particular edition.
+
+**Edition**: A numbered version of a document that has been, or is
+expected to be, published on GOV.UK. It is associated with a revision
+and a status. It is mutable and is a place where any edition level
+database constraints can be placed, such as only one live edition can exist for
+a document. It is possible that two editions of the same document share the
+same revision, this is in the case that they have identical content.
+
+**Revision**: Represents an immutable snapshot of the content of a document at a
+particular point in time. It has a number to indicate which revision of the
+document it is and stores who created it. Any request by a user that changes
+content should result in a single new revision, and only data that is
+changeable by a user directly should be stored as part of a revision. This is
+to directly map the concept of a revision to each time a user consciously
+revises a document. The [anatomy of a Revision model](#breakdown-of-revision)
+is explored further in this document.
+
+**Status**: Represents a state that an edition can hold such as: "draft" or
+"submitted for review". This model is coupled to the concept of status that is
+shown and changed by a user. Each time a user changes the status of an edition
+a new Status model is created and the user who created it stored. An edition
+can only have one status at any one time. It is intended to associate with
+specialist data for a state such as a Removal or Withdrawal, which allows the
+history of these to no longer be the responsibility of TimelineEntry. Initially
+this object is intended to be immutable however this may be changed if status
+changes become asynchronous operations, which may be the case with scheduling.
+
+### Approach to mutability/immutability
+
+A number of the models in Content Publisher are defined as immutable, most
+significantly Revision and its associated types. These models should be
+persisted to the database once and never updated or deleted, any need to change
+them requires creating a new record. This allows us to store a full history
+by only appending to the database.
+
+Accessing immutable models is intended to always be done by a foreign key and
```

Comment:
Could we think about rephrasing this paragraph, so it explains the problem before the solution? The wording is also a bit strange, as it suggests we 'invented' the Edition model to solve this particular problem, whereas I think it's intended for illustration purposes.